File: lib/python/opentoken/src/main/opentoken/tokens/definitions/t4_token.py
"""
Copyright (c) Truveta. All rights reserved.
"""

from typing import List
from opentoken.attributes.attribute_expression import AttributeExpression
from opentoken.attributes.person.birth_date_attribute import BirthDateAttribute
from opentoken.attributes.person.sex_attribute import SexAttribute
from opentoken.attributes.person.social_security_number_attribute import SocialSecurityNumberAttribute
from opentoken.tokens.token import Token


class T4Token(Token):
    """
    Represents the token definition for token T4.

    It is a collection of attribute expressions that are concatenated together
    to get the token signature. The token signature is as follows:
    social-security-number|U(gender)|birth-date
    """

    ID = "T4"

    def __init__(self):
        """Initialize the T4 token definition."""
        self._definition = [
            AttributeExpression(SocialSecurityNumberAttribute, "T|M(\\d+)"),
            AttributeExpression(SexAttribute, "T|U"),
            AttributeExpression(BirthDateAttribute, "T|D")
        ]

    def get_identifier(self) -> str:
        """Get the unique identifier for the token."""
        return self.ID

    def get_definition(self) -> List[AttributeExpression]:
        """Get the list of attribute expressions that define the token."""
        return self._definition

File: lib/java/opentoken/src/main/java/com/truveta/opentoken/tokens/definitions/T4Token.java
/**
 * Copyright (c) Truveta. All rights reserved.
 */
package com.truveta.opentoken.tokens.definitions;

import java.util.ArrayList;
import com.truveta.opentoken.attributes.AttributeExpression;
import com.truveta.opentoken.attributes.person.BirthDateAttribute;
import com.truveta.opentoken.attributes.person.SexAttribute;
import com.truveta.opentoken.attributes.person.SocialSecurityNumberAttribute;
import com.truveta.opentoken.tokens.Token;

/**
 * Represents the token definition for token T4.
 * 
 * <p>
 * It is a collection of attribute expressions
 * that are concatenated together to get the token signature.
 * The token signature is as follows:
 * social-security-number|U(gender)|birth-date
 * </p>
 * 
 * @see com.truveta.opentoken.tokens.Token Token
 */
public class T4Token implements Token {
    private static final String ID = "T4";

    private final ArrayList<AttributeExpression> definition = new ArrayList<>();

    public T4Token() {
        definition.add(new AttributeExpression(SocialSecurityNumberAttribute.class, "T|M(\\d+)"));
        definition.add(new AttributeExpression(SexAttribute.class, "T|U"));
        definition.add(new AttributeExpression(BirthDateAttribute.class, "T|D"));
    }

    @Override
    public String getIdentifier() {
        return ID;
    }

    @Override
    public ArrayList<AttributeExpression> getDefinition() {
        return definition;
    }
}
